libp2p/Session.cpp

```diff
@@ -298,14 +291,32 @@ void Session::drop(DisconnectReason _reason)
 
 void Session::disconnect(DisconnectReason _reason)
 {
+    if (m_dropped)
+        return;
+
     clog(VerbosityTrace, "p2pcap") << "Disconnecting (our reason: " << reasonOf(_reason) << ") from " << m_logSuffix;
 
-    if (m_socket->ref().is_open())
-    {
-        RLPStream s;
-        prep(s, DisconnectPacket, 1) << (int)_reason;
-        sealAndSend(s);
-    }
+    RLPStream s;
+    prep(s, DisconnectPacket, 1) << (int)_reason;
+    sealAndSend(s);
+    auto disconnectTimer = m_server->createTimer();
```

Comment:
I think this timer will be destroyed right after this function ends, and outstanding wait operation will be cancelled. (Probably should be made a member?)

Reply:
@gumb0 Agreed, in my prototype I had captured the timer in the handler but making it a member works as well 😄